prime-router/src/main/kotlin/fhirengine/translation/hl7/FhirTransformer.kt

```diff
@@ -0,0 +1,225 @@
+ * the schema validation process.
+ * @property constantSubstitutor the constant substitutor. Should be a static instance, but is not thread safe
+ */
+class FhirTransformer(
+    private val schemaRef: FhirTransformSchema,
+    private val strict: Boolean = false,
+    // the constant substitutor is not thread safe, so we need one instance per converter instead of using a shared copy
+    private val constantSubstitutor: ConstantSubstitutor = ConstantSubstitutor()
+) : ConfigSchemaProcessor() {
+    /**
+     * Transform a FHIR bundle based on the [schema] in the [schemaFolder] location.
+     * The transformer will error out if [strict] is set to true and there is an error during the translation. If [strict]
+     * is set to false (the default) then any translation errors are logged as a warning. Note [strict] does not affect
+     * the schema validation process.
+     */
+    constructor(
+        schema: String,
+        schemaFolder: String,
+        strict: Boolean = false,
+    ) : this(
+        schemaRef = fhirTransformSchemaFromFile(schema, schemaFolder),
+        strict = strict,
+    )
+
+    /**
+     * Transform a FHIR bundle based on the [schema] (which includes a folder location).
+     * The transformer will error out if [strict] is set to true and there is an error during the translation. If [strict]
+     * is set to false (the default) then any translation errors are logged as a warning. Note [strict] does not affect
+     * the schema validation process.
+     */
+    constructor(
+        schema: String,
+        strict: Boolean = false,
+    ) : this(
+        schemaRef = fhirTransformSchemaFromFile(
+            FilenameUtils.getName(schema),
+            FilenameUtils.getPathNoEndSeparator(schema)
+        ),
+        strict = strict,
+    )
+
+    /**
+     * Transform the given [bundle]. The bundle passed in will be updated directly, and will also be returned.
+     * @return the transformed bundle
+     */
+    fun transform(bundle: Bundle): Bundle {
+        val dupes = schemaRef.duplicateElements
+        if (dupes.isNotEmpty()) { // value is the number of matches
+            throw SchemaException("Schema ${schemaRef.name} has multiple elements with the same name: ${dupes.keys}")
+        }
+        transformWithSchema(schemaRef, bundle = bundle, focusResource = bundle)
+        return bundle
+    }
+
+    /**
+     * Transform the [bundle] using the elements in the given [schema] using [context] starting at the
+     * [focusResource] in the bundle. Set [debug] to true to enable debug statements to the logs.
+     */
+    private fun transformWithSchema(
+        schema: FhirTransformSchema,
+        bundle: Bundle,
+        focusResource: Base,
+        context: CustomContext = CustomContext(bundle, focusResource),
+        debug: Boolean = false
+    ) {
+        val logLevel = if (debug) Level.INFO else Level.DEBUG
```

Comment:
Does this need to be flipped where it is Level.DEBUG if debug is true?

Reply:
Good question, what this is actually doing is setting the level for the log messages, so if `debug` is true, then we're logging things at the informational level so they will be more visible. If `debug` is false, then we log things at the debug level so they'll be less visible. I'm not super familiar with our logger, but that's my understanding of it, and this was following the lead from existing code in `FhirToHl7Converter.kt`.